doc/developer/design/20240502_deprecate_subsources.md

```diff
@@ -0,0 +1,190 @@
+# sql+storage: deprecate concept of subsources
+
+- Associated:
+    - [storage/sources: make (sub)sources a flavor of tables #20208
+      ](https://github.com/MaterializeInc/materialize/issues/20208)
+    - [sql/storage: consider if TEXT COLUMNS, IGNORE COLUMNS belongs on
+      subsources
+      #26774](https://github.com/MaterializeInc/materialize/issues/26774)
+
+## The Problem
+
+There are two distinct notions of `SUBSOURCE`s in MZ, both of which can be
+folded neatly into other concepts.
+
+- Ingestion export subsources should be tables (as should the current primary
+  source's collection)
+- Progress subsources should become the relation returned when querying sources
+
+Doing both of these lets us remove the concept of subsources from Materialize
+entirely. It, by definition, also lets us get rid of the notion of "progress
+subsources," which is a concept that users often find confusion.
+
+## Success Criteria
+
+- Users no longer need to use the keyword `SUBSOURCE` to interact with
+  Materialize in any capacity; and we do this without deprecating any
+  functionality whatsoever.
+
+## Solution Proposal
+
+### SOURCEs should output to TABLES
+
+In the current version of Materialize, we have an impedance mismatch between
+single-output and multi-output sources.
+
+- Single-ouput sources write their data directly to the `SOURCE` itself, i.e.
+  users query the `SOURCE` object to read the data.
+- Multi-output sources write their data to `SUBSOURCES`. The primary source to
+  which they belong is an empty, unused (though still queryable) collection.
+
+I propose that the output of all sources, both single- and multi-output, should
+be to `TABLE`s (albeit special, read-only tables). Users already have an
+intuition around how `TABLE`s work, and understanding "read-only" tables is much
+simpler than doubling down on subsources––a concept we invented.
+
+#### Why not just rename `SUBSOURCE` to `TABLE`?
+
+This leaves the impedance mismatch in place where users have to `SELECT` from
+the source if they're using e.g. Kafka, but have to find the right `TABLE` to
+`SELECT` from for PG sources.
+
+### SOURCES should contain their progress information
+
+The other type of `SUBSOURCE` Materialize supports is known as the
+remap/progress collection. While most users don't access this data, it's an
+important debugging tool.
+
+I propose that the progress collection should be moved "into" the source, i.e.
+querying a source should return the what-is-today-the-progress-collection's
+data.
+
+### New API
+
+The API to support this is similar to the existing `CREATE SOURCE` API. However,
+`ALTER SOURCE...ADD SUBSOURCE` will disappear and be replaced with a flavor of
+`CREATE TABLE`.
+
+(h/t @benesch for this)
+
+***
+
+Create a Kafka source that reads from a topic named `t`:
+
+```sql
+CREATE SOURCE kafka_src
+FROM KAFKA CONNECTION kafka_conn (TOPIC 't')
+FOR ALL TABLES
```

Comment:
From a chat with @benesch: does this imply that there would be a `FOR NO TABLES` (or similar) option that only creates the source, but suppresses automatic table creation? This would be useful so we can decouple source management and table creation in dbt.

Reply:
Also curious if that feature already exists today as `FOR TABLES ()`.

Reply:
Today, you'd get errors for not having created any objects in `CREATE SOURCE`. However, we could introduce support for `FOR TABLES ()`. I dont know exactly what would happen w/r/t frontier management w/ PG and MySQL sources, but it's something we could figure out.